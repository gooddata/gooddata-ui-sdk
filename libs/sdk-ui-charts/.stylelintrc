{
  "extends": [
    "../../.stylelintrc"
  ],
  "rules": {
    "max-nesting-depth": 4
  }
}
