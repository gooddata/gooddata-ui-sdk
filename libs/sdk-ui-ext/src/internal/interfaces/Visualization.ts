// (C) 2019-2021 GoodData Corporation
import { ISeparators } from "@gooddata/numberjs";
import isEmpty from "lodash/isEmpty";
import { IAnalyticalBackend, IExecutionFactory, ISettings, ITheme } from "@gooddata/sdk-backend-spi";
import {
    IColorPalette,
    Identifier,
    IInsight,
    IInsightDefinition,
    ITotal,
    ObjRef,
    VisualizationProperties,
    IExecutionConfig,
} from "@gooddata/sdk-model";
import {
    ChartType,
    ErrorCodes,
    GoodDataSdkError,
    IDrillableItem,
    IDrillEvent,
    IHeaderPredicate,
    ILocale,
    IPushData,
    IVisualizationCallbacks,
    OverTimeComparisonType,
    SdkErrorType,
    VisualizationEnvironment,
} from "@gooddata/sdk-ui";

export type RenderFunction = (component: any, target: Element) => void;

export interface IVisConstruct {
    backend: IAnalyticalBackend;
    // TODO: rename to workspace
    projectId: string;
    element: string;
    configPanelElement: string;

    callbacks: IVisCallbacks;
    environment?: VisualizationEnvironment;
    locale?: ILocale;
    // TODO: rename to settings
    featureFlags?: ISettings;
    visualizationProperties: VisualizationProperties;
    renderFun: RenderFunction;
}

export interface ICustomProps {
    drillableItems?: Array<IDrillableItem | IHeaderPredicate>;
    totalsEditAllowed?: boolean;
}

export interface IDimensions {
    width?: number; // Note: will not be optional once we start sending it
    height: number;
}

/**
 * @alpha
 */
export interface IVisProps {
    dimensions?: IDimensions;
    custom?: ICustomProps;
    locale?: ILocale;
    dateFormat?: string;
    config?: IGdcConfig;
    executionConfig?: IExecutionConfig;
    theme?: ITheme;

    /*
     * This can be used to override configuration of visualization when embedding through InsightView. The pluggable
     * visualization implementation MUST take these custom properties and override the the effective visualization
     * config with them.
     *
     * It is important that the override happens just before sending the config to the underlying chart -
     * the custom config passed through here will be in the format that is understood by the underlying chart.
     */
    customVisualizationConfig?: any;
}

export interface IVisualizationOptions {
    dateOptionsDisabled: boolean;
}

export interface IVisCallbacks extends IVisualizationCallbacks {
    pushData(data: IPushData, options?: IVisualizationOptions): void;
}

export interface IBucketFilterElement {
    title: string;
    uri: string;
}

export type ComparisonConditionOperator =
    | "GREATER_THAN"
    | "GREATER_THAN_OR_EQUAL_TO"
    | "LESS_THAN"
    | "LESS_THAN_OR_EQUAL_TO"
    | "EQUAL_TO"
    | "NOT_EQUAL_TO";

export interface IComparisonCondition {
    readonly comparison: {
        readonly operator: ComparisonConditionOperator;
        readonly value: number;
    };
}

export type RangeConditionOperator = "BETWEEN" | "NOT_BETWEEN";

export interface IRangeCondition {
    readonly range: {
        readonly operator: RangeConditionOperator;
        readonly from: number;
        readonly to: number;
    };
}

export type IMeasureValueFilterCondition = IComparisonCondition | IRangeCondition;

export interface IBucketFilterInterval {
    granularity: string;
    interval: string[];
    name: string;
}

export interface IAttributeFilter {
    attribute: string;
    isInverted: boolean;
    totalElementsCount: number;
    displayFormRef: ObjRef;
    selectedElements: Array<{
        title: string;
        uri: string;
    }>;
}

export const DATE_DATASET_ATTRIBUTE = "attr.datedataset";

export interface IDateFilter {
    attribute: "attr.datedataset";
    overTimeComparisonType: OverTimeComparisonType;
    interval: {
        granularity: string;
        interval: [string, string] | [number, number];
        name: string;
        type: "relative" | "absolute";
    };
}

export interface IMeasureValueFilter {
    measureLocalIdentifier: string;
    condition?: IMeasureValueFilterCondition;
}

export type RankingFilterOperator = "TOP" | "BOTTOM";

export interface IRankingFilter {
    measure: string;
    attributes?: string[];
    operator: RankingFilterOperator;
    value: number;
}

export type IBucketFilter = IAttributeFilter | IDateFilter | IMeasureValueFilter | IRankingFilter;

export interface ISort {
    direction: "asc" | "desc";
}

export interface IBucketItem {
    localIdentifier: string;
    type?: string;
    aggregation?: boolean;
    attribute?: string;
    filters?: IBucketFilter[];
    format?: string;
    granularity?: string;
    showInPercent?: boolean;
    showOnSecondaryAxis?: boolean;
    sort?: ISort;
    masterLocalIdentifier?: string;
    overTimeComparisonType?: OverTimeComparisonType;
    operandLocalIdentifiers?: Array<string | null> | null;
    operator?: string | null;

    dfRef?: ObjRef;
    locationDisplayFormRef?: ObjRef;
    dateDatasetRef?: ObjRef;
}

export interface IFiltersBucketItem extends IBucketItem {
    autoCreated?: boolean;
}

// TODO: SDK8: rename this :) the original name IBucket conflicted with what we have in model;
// this interface and all the other similar interface (bucket item, filters etc) are specifically used
// in the reference point
export interface IBucketOfFun {
    localIdentifier: string;
    items: IBucketItem[];
    totals?: ITotal[];
    chartType?: string;
}

export interface IFilters {
    localIdentifier: "filters";
    items: IFiltersBucketItem[];
}

export interface IRecommendations {
    [key: string]: boolean;
}

export interface IBucketUiConfig {
    accepts?: string[];
    canAddItems?: boolean;
    warningMessage?: string;
    title?: string;
    subtitle?: string;
    icon?: string;
    allowsDuplicateItems?: boolean;

    // allow more than one date item in the bucket, regardless of date dimension
    allowsDuplicateDates?: boolean;
    allowsReordering?: boolean;
    allowsSwapping?: boolean;
    enabled?: boolean;
    itemsLimit?: number;
    itemsLimitByType?: {
        date?: number;
        metric?: number;
        fact?: number;
        attribute?: number;
    };
    isShowInPercentEnabled?: boolean;
    isShowInPercentVisible?: boolean;
    isShowOnSecondaryAxisVisible?: boolean;
    allowShowOnSecondaryAxis?: boolean;
    allowSelectChartType?: boolean;
    allowOptionalStacking?: boolean;

    // indicates that the visualization prefers date items in the bucket to have the same dimension
    preferSynchronizedDates?: boolean;
}

export interface IBucketsUiConfig {
    [localIdentifier: string]: IBucketUiConfig;
}

export interface IExportUiConfig {
    supported?: boolean;
}

export interface INoMetricUiConfig {
    supported?: boolean;
}

export interface IOpenAsReportUiConfig {
    supported?: boolean;
    warningMessage?: string;
}

export interface ICustomError {
    heading: string;
    text: string;
}

export interface IOptionalStacking {
    supported?: boolean;
    disabled?: boolean;
    stackMeasures?: boolean;
    stackMeasuresToPercent?: boolean;
    canStackInPercent?: boolean;
}

export interface ISupportedLocationIcon {
    supported?: boolean;
}

export interface IUiConfig {
    buckets: IBucketsUiConfig;
    recommendations?: IRecommendations;
    exportConfig?: IExportUiConfig;
    noMetricAccepted?: INoMetricUiConfig;
    openAsReport?: IOpenAsReportUiConfig;
    customError?: ICustomError;
    supportedOverTimeComparisonTypes?: OverTimeComparisonType[];
    supportedChartTypes?: ChartType[];
    axis?: string;
    optionalStacking?: IOptionalStacking;
    supportedLocationIcon?: ISupportedLocationIcon;
}

export interface IVisualizationProperties {
    // This can be anything depending on a visualization type
    // perhaps consider adding: sortItems?: AFM.SortItem[]
    [property: string]: any;
}

export interface IReferencePoint {
    buckets: IBucketOfFun[];
    filters: IFilters;
    properties?: IVisualizationProperties; // properties are under plugvis creator's control
}

export interface IReferences {
    [identifier: string]: string;
}

export interface IExtendedReferencePoint {
    buckets: IBucketOfFun[];
    filters: IFilters;
    properties?: IVisualizationProperties; // properties are under plugvis creator's control
    uiConfig: IUiConfig;
}

/**
 * @alpha
 */
export interface IVisualization {
    /**
     * Update and re-render visualization to reflect change of insight state.
     *
     * Currently it is possible that context (AD/KD) will send insight that is not in a valid state:
     *
     * -  insight might be empty
     * -  insight might not be completely defined (user did not yet specify measures etc)
     *
     * It is the responsibility of the implementation to verify the state of insight during the update and
     * if there is anything amiss communicate with the context using the onError callback which it
     * received during construction time via {@link IVisConstruct}
     *
     * The loading state of the visualization must be communicated using the onLoadingChanged callback which
     * is also passed during construction time.
     *
     * @param props - some runtime properties
     * @param insight - new state of insight
     * @param insightPropertiesMeta - new state of insight properties meta
     * @param executionFactory - execution factory to use when triggering calculation on backend
     */
    update(
        props: IVisProps,
        insight: IInsightDefinition,
        insightPropertiesMeta: any,
        executionFactory: IExecutionFactory,
    ): void;

    unmount(): void;

    addNewDerivedBucketItems(
        referencePoint: IReferencePoint,
        newDerivedBucketItems: IBucketItem[],
    ): Promise<IReferencePoint>;

    /**
     * Called every time the reference point or the visualization class change
     * to allow visualizations to get updated ExtendedReferencePoint.
     * @param referencePoint The new value of the reference point.
     * @param previousReferencePoint The previous value of the reference point.
     * This value is only provided if the visualization class was not changed
     * (i. e. both points are related to the same visualization class).
     * @returns Promise of the new ExtendedReferencePoint.
     */
    getExtendedReferencePoint(
        referencePoint: IReferencePoint,
        previousReferencePoint?: IReferencePoint,
    ): Promise<IExtendedReferencePoint>;

    /**
     * Called when the Drill Down is performed, used to get the Drill Down target {@link IInsight} instance.
     *
     * The exact contract depends on individual {@link IInsight} type, but generally it should replace
     * the drilled attribute with the Drill Down target target attribute and include the filters from the
     * drill event into the returned {@link IInsight}.
     *
     * @param source {@link IInsight} to be used for the the Drill Down execution
     * @param drillDownContext Drill Down configuration used to properly create the result
     * @returns {@code source} as the Drill Down target {@link IInsight}
     */
    getInsightWithDrillDownApplied(source: IInsight, drillDownContext: IDrillDownContext): IInsight;
}

export interface IGdcConfig {
    separators?: ISeparators;
    colorPalette?: IColorPalette;
    isExportMode?: boolean;
    isInEditMode?: boolean;
    mapboxToken?: string;
    maxWidth?: number;
    maxHeight?: number;
    forceDisableDrillOnAxes?: boolean;
}

/**
 * Class name of element where pluggable visualization is supposed to render its configuration
 * panels.
 *
 * @alpha
 */
export const ConfigPanelClassName = "gd-configuration-panel-content";

/**
 * @alpha
 */
export const PluggableVisualizationErrorCodes = {
    /**
     * If pluggable visualization is asked to render itself but its buckets do not contain the right 'stuff',
     * then this is the error code to communicate the fact.
     */
    INVALID_BUCKETS: "INVALID_BUCKETS",

    /**
     * This error means that empty AFM was went to the GoodData.UI and as such can't be executed.
     */
    EMPTY_AFM: "EMPTY_AFM",
};

/**
 * @alpha
 */
export type PluggableVisualizationErrorType = keyof typeof PluggableVisualizationErrorCodes;

/**
 * @alpha
 */
export class InvalidBucketsSdkError extends GoodDataSdkError {
    public readonly pveType: PluggableVisualizationErrorType;

    constructor(cause?: Error) {
        super(ErrorCodes.UNKNOWN_ERROR as SdkErrorType, undefined, cause);

        this.pveType = "INVALID_BUCKETS";
    }

    public getErrorCode(): string {
        return this.pveType;
    }
}

/**
 * @alpha
 */
export class EmptyAfmSdkError extends GoodDataSdkError {
    public readonly pveType: PluggableVisualizationErrorType;

    constructor(cause?: Error) {
        super(ErrorCodes.UNKNOWN_ERROR as SdkErrorType, undefined, cause);

        this.pveType = "EMPTY_AFM";
    }

    public getErrorCode(): string {
        return this.pveType;
    }
}

/**
 * @alpha
 */
export type PluggableVisualizationError = InvalidBucketsSdkError | EmptyAfmSdkError;

/**
 * @alpha
 */
export function isPluggableVisualizationError(obj: unknown): obj is PluggableVisualizationError {
    return !isEmpty(obj) && (obj as any).pveType !== undefined;
}

/**
 * @alpha
 */
export function isInvalidBuckets(obj: unknown): obj is InvalidBucketsSdkError {
    return !isEmpty(obj) && (obj as InvalidBucketsSdkError).pveType === "INVALID_BUCKETS";
}

/**
 * @alpha
 */
export function isEmptyAfm(obj: unknown): obj is EmptyAfmSdkError {
    return !isEmpty(obj) && (obj as EmptyAfmSdkError).pveType === "EMPTY_AFM";
}

/**
 * Source implicit drill down attribute local Identifier
 *
 * @alpha
 */
export interface IDrillFromAttribute {
    drillFromAttribute: {
        localIdentifier: Identifier;
    };
}

/**
 * Target implicit drill down attribute display form
 *
 * @alpha
 */
export interface IDrillToAttribute {
    drillToAttribute: {
        attributeDisplayForm: ObjRef;
    };
}

/**
 * Implicit drill down definition
 *
 * @alpha
 */
export interface IImplicitDrillDown {
    implicitDrillDown: {
        from: IDrillFromAttribute;
        drillDownStep: IDrillToAttribute;
    };
}

/**
 * Implicit drill down context
 *
 * @alpha
 */
export interface IDrillDownContext {
    drillDefinition: IImplicitDrillDown;
    event: IDrillEvent;
}
