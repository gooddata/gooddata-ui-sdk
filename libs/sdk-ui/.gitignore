styles/css
styles/internal/css

src/base/localization/bundles/*.ts
