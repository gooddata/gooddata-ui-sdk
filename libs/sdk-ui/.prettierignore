#we ignore these files because they are generated after build
src/base/localization/bundles/*.ts
