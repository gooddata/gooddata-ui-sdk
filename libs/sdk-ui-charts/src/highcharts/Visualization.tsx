// (C) 2007-2018 GoodData Corporation
import { IDataView } from "@gooddata/sdk-backend-spi";
import * as invariant from "invariant";
import * as React from "react";
import { IDrillableItem, OnFiredDrillEvent, IHeaderPredicate } from "@gooddata/sdk-ui";
import ChartTransformation, { renderHighCharts } from "./chart/ChartTransformation";
import Highcharts from "./chart/highcharts/highchartsEntryPoint";
import { IChartConfig } from "./Config";

import { isChartSupported, stringifyChartTypes } from "./utils/common";
import isEqual = require("lodash/isEqual");
import isFunction = require("lodash/isFunction");
import noop = require("lodash/noop");
import omitBy = require("lodash/omitBy");

export interface IVisualizationProps {
    height: number;
    width: number;
    config: IChartConfig;
    numericSymbols?: string[];

    dataView: IDataView;
    drillableItems: Array<IDrillableItem | IHeaderPredicate>;
    locale?: string;

    onDrill?: OnFiredDrillEvent;
    afterRender?: () => void;
    onDataTooLarge(): void;
    onNegativeValues(): void;
    onLegendReady(): void;
    pushData?(data: any): void;
}

export class Visualization extends React.Component<IVisualizationProps> {
    public static defaultProps = {
        locale: "en-US",
        numericSymbols: [] as string[],
        onDrill: () => true,
        afterRender: noop,
    };

    constructor(props: IVisualizationProps) {
        super(props);

        this.setNumericSymbols(this.props);
    }

    public UNSAFE_componentWillReceiveProps(nextProps: IVisualizationProps) {
        this.setNumericSymbols(nextProps);
    }

    public shouldComponentUpdate(nextProps: IVisualizationProps) {
        return !isEqual(omitBy(this.props, isFunction), omitBy(nextProps, isFunction));
    }

    public setNumericSymbols(props: IVisualizationProps) {
        const { numericSymbols } = props;

        if (numericSymbols && numericSymbols.length) {
            Highcharts.setOptions({
                lang: {
                    numericSymbols,
                },
            });
        }
    }

    public render(): JSX.Element {
        const visType = this.props.config.type;

        if (isChartSupported(visType)) {
            const {
                height,
                width,
                config,
                dataView,
                drillableItems,
                onDrill,
                afterRender,
                onDataTooLarge,
                onNegativeValues,
                onLegendReady,
                locale,
                pushData,
            } = this.props;

            return (
                <ChartTransformation
                    height={height}
                    width={width}
                    config={config}
                    drillableItems={drillableItems}
                    locale={locale}
                    dataView={dataView}
                    afterRender={afterRender}
                    onDrill={onDrill}
                    onDataTooLarge={onDataTooLarge}
                    onNegativeValues={onNegativeValues}
                    onLegendReady={onLegendReady}
                    pushData={pushData}
                    renderer={renderHighCharts}
                />
            );
        }

        invariant(
            false,
            `Unknown visualization type: ${visType}. Supported visualization types: ${stringifyChartTypes()}`,
        );

        return null;
    }
}
