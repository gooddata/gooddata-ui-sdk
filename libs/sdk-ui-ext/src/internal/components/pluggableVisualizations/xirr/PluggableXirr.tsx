// (C) 2019-2022 GoodData Corporation

import { IExecutionFactory, ISettings } from "@gooddata/sdk-backend-spi";
import {
    attributeLocalId,
    bucketAttributes,
    IDimension,
    IInsightDefinition,
    insightBucket,
    MeasureGroupIdentifier,
    newDimension,
} from "@gooddata/sdk-model";
import { BucketNames, IPushData } from "@gooddata/sdk-ui";

import { CoreXirr, updateConfigWithSettings } from "@gooddata/sdk-ui-charts";
import React from "react";
import { render } from "react-dom";
import {
    IExtendedReferencePoint,
    IReferencePoint,
    IVisConstruct,
    IVisProps,
    IVisualizationOptions,
    RenderFunction,
} from "../../../interfaces/Visualization";
import {
    removeAllArithmeticMeasuresFromDerived,
    removeAllDerivedMeasures,
    sanitizeFilters,
} from "../../../utils/bucketHelper";

import { hasGlobalDateFilter } from "../../../utils/bucketRules";
import { unmountComponentsAtNodes } from "../../../utils/domHelper";
import {
    getReferencePointWithSupportedProperties,
    getSupportedProperties,
} from "../../../utils/propertiesHelper";
import { removeSort } from "../../../utils/sort";
import { getDefaultXirrUiConfig, getXirrUiConfig } from "../../../utils/uiConfigHelpers/xirrUiConfigHelper";

import UnsupportedConfigurationPanel from "../../configurationPanels/UnsupportedConfigurationPanel";
import { AbstractPluggableVisualization } from "../AbstractPluggableVisualization";
import { getXirrBuckets } from "./xirrBucketHelper";
import cloneDeep from "lodash/cloneDeep";

export class PluggableXirr extends AbstractPluggableVisualization {
    private settings?: ISettings;
    private renderFun: RenderFunction;

    constructor(props: IVisConstruct) {
        super(props);

        this.settings = props.featureFlags;
        this.renderFun = props.renderFun;
    }

    public unmount(): void {
        unmountComponentsAtNodes([this.element, this.configPanelElement]);
    }

    public getExtendedReferencePoint = async (
        referencePoint: Readonly<IReferencePoint>,
    ): Promise<IExtendedReferencePoint> => {
        const referencePointCloned = cloneDeep(referencePoint);
        let newReferencePoint: IExtendedReferencePoint = {
            ...referencePointCloned,
            uiConfig: getDefaultXirrUiConfig(),
        };

        if (!hasGlobalDateFilter(referencePoint.filters)) {
            newReferencePoint = removeAllArithmeticMeasuresFromDerived(newReferencePoint);
            newReferencePoint = removeAllDerivedMeasures(newReferencePoint);
        }

        const buckets = getXirrBuckets(referencePoint);
        newReferencePoint.buckets = buckets;

        newReferencePoint.uiConfig = getXirrUiConfig(newReferencePoint, this.intl);

        newReferencePoint = removeSort(newReferencePoint);
        newReferencePoint = getReferencePointWithSupportedProperties(
            newReferencePoint,
            this.supportedPropertiesList,
        );

        return sanitizeFilters(newReferencePoint);
    };

    public getExecution(
        options: IVisProps,
        insight: IInsightDefinition,
        executionFactory: IExecutionFactory,
    ) {
        const { dateFormat } = options;

        return executionFactory
            .forInsight(insight)
            .withDimensions(...this.getXirrDimensions(insight))
            .withDateFormat(dateFormat);
    }

    protected renderVisualization(
        options: IVisProps,
        insight: IInsightDefinition,
        executionFactory: IExecutionFactory,
    ): void {
        const { locale, custom = {}, config } = options;
        const { drillableItems } = custom;
        const execution = this.getExecution(options, insight, executionFactory);

        this.renderFun(
            <CoreXirr
                execution={execution}
                drillableItems={drillableItems}
                onDrill={this.onDrill}
                locale={locale}
                config={updateConfigWithSettings(config, this.settings)}
                afterRender={this.afterRender}
                onLoadingChanged={this.onLoadingChanged}
                pushData={this.pushData}
                onError={this.onError}
                LoadingComponent={null}
                ErrorComponent={null}
            />,
            document.querySelector(this.element),
        );
    }

    protected renderConfigurationPanel(): void {
        if (document.querySelector(this.configPanelElement)) {
            const properties = this.visualizationProperties ?? {};

            render(
                <UnsupportedConfigurationPanel
                    locale={this.locale}
                    pushData={this.pushData}
                    properties={getSupportedProperties(properties, this.supportedPropertiesList)}
                />,
                document.querySelector(this.configPanelElement),
            );
        }
    }

    private getXirrDimensions(insight: IInsightDefinition): IDimension[] {
        const attribute = insightBucket(insight, BucketNames.ATTRIBUTE);

        if (attribute && attribute.items.length) {
            return [
                newDimension([MeasureGroupIdentifier, ...bucketAttributes(attribute).map(attributeLocalId)]),
            ];
        }

        return [newDimension([MeasureGroupIdentifier])];
    }

    private withEmptyAttributeTargets(data: IPushData): IPushData {
        return {
            ...data,
            availableDrillTargets: {
                ...data?.availableDrillTargets,
                attributes: [],
            },
        };
    }

    // This is effectively calling super.pushData()
    // https://stackoverflow.com/questions/31088947/inheritance-method-call-triggers-typescript-compiler-error
    // https://github.com/basarat/typescript-book/blob/master/docs/arrow-functions.md#tip-arrow-functions-and-inheritance
    private superPushData = this.pushData;

    protected pushData = (data: IPushData, options?: IVisualizationOptions): void => {
        // For xirr chart we do not support drilling from attributes.
        const filterAtrributes = this.withEmptyAttributeTargets(data);
        this.superPushData(filterAtrributes, options);
    };
}
