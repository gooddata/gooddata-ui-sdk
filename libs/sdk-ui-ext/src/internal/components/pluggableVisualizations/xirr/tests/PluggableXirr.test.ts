// (C) 2019-2024 GoodData Corporation
import { dummyBackend } from "@gooddata/sdk-backend-mockingbird";

import { PluggableXirr } from "../PluggableXirr.js";
import { IVisConstruct, IVisProps } from "../../../../interfaces/Visualization.js";
import * as referencePointMocks from "../../../../tests/mocks/referencePointMocks.js";
import * as testMocks from "../../../../tests/mocks/testMocks.js";
import { IDrillableItem } from "@gooddata/sdk-ui";
import { CoreXirr, ICoreChartProps } from "@gooddata/sdk-ui-charts";
import { getLastRenderEl } from "../../tests/testHelpers.js";
import { describe, it, expect, vi, afterEach } from "vitest";

describe("PluggableXirr", () => {
    const mockElement = document.createElement("div");
    const mockConfigElement = document.createElement("div");
    const mockRenderFun = vi.fn();
    const executionFactory = dummyBackend().workspace("PROJECTID").execution();
    const defaultProps: IVisConstruct = {
        projectId: "PROJECTID",
        backend: dummyBackend(),
        element: () => mockElement,
        configPanelElement: () => mockConfigElement,
        visualizationProperties: {},
        callbacks: {
            afterRender: vi.fn(),
            pushData: vi.fn(),
            onLoadingChanged: vi.fn(),
            onError: vi.fn(),
        },
        renderFun: mockRenderFun,
        unmountFun: vi.fn(),
    };

    afterEach(() => {
        mockRenderFun.mockReset();
    });

    function createComponent(customProps: Partial<IVisConstruct> = {}) {
        return new PluggableXirr({
            ...defaultProps,
            ...customProps,
        });
    }

    describe("init", () => {
        it("should not call pushData during init", () => {
            const pushData = vi.fn();

            createComponent({
                callbacks: {
                    pushData,
                },
            });

            expect(pushData).not.toHaveBeenCalled();
        });
    });

    describe("update", () => {
        const emptyPropertiesMeta = {};

        function getTestOptions(): IVisProps {
            const drillableItems: IDrillableItem[] = [];
            return {
                dimensions: {
                    width: 12,
                    height: 14,
                },
                custom: {
                    drillableItems,
                },
                locale: "en-US",
            };
        }

        it("should not render xirr when dataSource is missing", () => {
            const xirr = createComponent();
            const options: IVisProps = getTestOptions();

            xirr.update({ ...options }, testMocks.emptyInsight, emptyPropertiesMeta, executionFactory);

            const renderEl = getLastRenderEl(mockRenderFun, mockElement);
            expect(renderEl).toBeUndefined();
        });

        it("should render XIRR by react to given element passing down properties", () => {
            const xirr = createComponent({
                ...defaultProps,
                renderFun: mockRenderFun,
                featureFlags: { enableKDWidgetCustomHeight: false },
            });
            const options: IVisProps = getTestOptions();

            xirr.update(options, testMocks.insightWithSingleMeasure, emptyPropertiesMeta, executionFactory);

            const renderEl = getLastRenderEl<ICoreChartProps>(mockRenderFun, mockElement);
            expect(renderEl.type).toBe(CoreXirr);
            expect(renderEl.props.config.enableCompactSize).toBeUndefined();
        });

        it("should render XIRR by react to given element passing down properties when FF enableKDWidgetCustomHeight is set to true", () => {
            const xirr = createComponent({
                ...defaultProps,
                renderFun: mockRenderFun,
                featureFlags: { enableKDWidgetCustomHeight: true },
            });
            const options: IVisProps = getTestOptions();

            xirr.update(options, testMocks.insightWithSingleMeasure, emptyPropertiesMeta, executionFactory);

            const renderEl = getLastRenderEl<ICoreChartProps>(mockRenderFun, mockElement);
            expect(renderEl.type).toBe(CoreXirr);
            expect(renderEl.props.config.enableCompactSize).toBe(true);
        });

        it("should correctly set config.disableDrillUnderline from FeatureFlag disableKpiDashboardHeadlineUnderline", () => {
            const xirr = createComponent({
                featureFlags: {
                    disableKpiDashboardHeadlineUnderline: true,
                },
                renderFun: mockRenderFun,
            });

            const options: IVisProps = getTestOptions();

            xirr.update(options, testMocks.insightWithSingleMeasure, emptyPropertiesMeta, executionFactory);

            const renderEl = getLastRenderEl<ICoreChartProps>(mockRenderFun, mockElement);
            expect(renderEl.type).toBe(CoreXirr);
            expect(renderEl.props.config.disableDrillUnderline).toBe(true);
        });
    });

    describe("getExtendedReferencePoint", () => {
        it("should return proper extended reference point", async () => {
            const extendedReferencePoint = await createComponent().getExtendedReferencePoint(
                referencePointMocks.measuresAndDateReferencePoint,
            );

            expect(extendedReferencePoint).toMatchSnapshot();
        });

        it("should correctly process empty reference point", async () => {
            const headline = createComponent();

            const extendedReferencePoint = await headline.getExtendedReferencePoint(
                referencePointMocks.emptyReferencePoint,
            );

            expect(extendedReferencePoint).toMatchSnapshot();
        });
    });

    describe("PluggableXirr renderVisualization and renderConfigurationPanel", () => {
        it("should mount on the element defined by the callback", () => {
            const visualization = createComponent();

            visualization.update({}, testMocks.insightWithSingleMeasure, {}, executionFactory);

            // 1st call for rendering element
            // 2nd call for rendering config panel
            expect(mockRenderFun).toHaveBeenCalledTimes(2);
            expect(getLastRenderEl(mockRenderFun, mockElement)).toBeDefined();
            expect(getLastRenderEl(mockRenderFun, mockConfigElement)).toBeDefined();
        });
    });
});
