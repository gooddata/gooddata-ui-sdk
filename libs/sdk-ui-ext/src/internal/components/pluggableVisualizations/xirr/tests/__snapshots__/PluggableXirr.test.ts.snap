// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PluggableXirr > getExtendedReferencePoint > should correctly process empty reference point 1`] = `
{
  "buckets": [
    {
      "items": [],
      "localIdentifier": "measures",
    },
    {
      "items": [],
      "localIdentifier": "attribute",
    },
  ],
  "filters": {
    "items": [],
    "localIdentifier": "filters",
  },
  "properties": {},
  "uiConfig": {
    "buckets": {
      "attribute": {
        "accepts": [
          "date",
        ],
        "allowsReordering": false,
        "allowsSwapping": true,
        "canAddItems": true,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "itemsLimit": 1,
        "itemsLimitByType": {
          "date": 1,
        },
        "title": "Date attribute",
      },
      "filters": {
        "accepts": [
          "attribute",
          "date",
        ],
        "allowsReordering": false,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "itemsLimit": 20,
        "itemsLimitByType": {
          "date": 1,
        },
      },
      "measures": {
        "accepts": [
          "metric",
          "fact",
        ],
        "allowsDuplicateItems": true,
        "allowsReordering": true,
        "allowsSwapping": true,
        "canAddItems": true,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "isShowInPercentVisible": false,
        "itemsLimit": 1,
        "title": "Metric",
      },
    },
    "customError": {
      "heading": "Incomplete configuration",
      "text": "Make sure you have selected one metric and one date attribute",
    },
    "exportConfig": {
      "supported": false,
    },
    "openAsReport": {
      "supported": false,
    },
    "recommendations": {
      "supportsFilteringRecommendation": true,
    },
    "supportedOverTimeComparisonTypes": [],
  },
}
`;

exports[`PluggableXirr > getExtendedReferencePoint > should return proper extended reference point 1`] = `
{
  "buckets": [
    {
      "items": [
        {
          "aggregation": null,
          "attribute": "aazb6kroa3iC",
          "localIdentifier": "m1",
          "showInPercent": null,
          "showOnSecondaryAxis": null,
          "type": "metric",
        },
      ],
      "localIdentifier": "measures",
    },
    {
      "items": [
        {
          "aggregation": null,
          "attribute": "attr.datedataset",
          "dateDatasetRef": {
            "identifier": "ds1",
          },
          "localIdentifier": "a1",
          "showInPercent": null,
          "type": "date",
        },
      ],
      "localIdentifier": "attribute",
    },
  ],
  "filters": {
    "items": [],
    "localIdentifier": "filters",
  },
  "properties": {},
  "uiConfig": {
    "buckets": {
      "attribute": {
        "accepts": [
          "date",
        ],
        "allowsReordering": false,
        "allowsSwapping": true,
        "canAddItems": false,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "itemsLimit": 1,
        "itemsLimitByType": {
          "date": 1,
        },
        "title": "Date attribute",
      },
      "filters": {
        "accepts": [
          "attribute",
          "date",
        ],
        "allowsReordering": false,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "itemsLimit": 20,
        "itemsLimitByType": {
          "date": 1,
        },
      },
      "measures": {
        "accepts": [
          "metric",
          "fact",
        ],
        "allowsDuplicateItems": true,
        "allowsReordering": true,
        "allowsSwapping": true,
        "canAddItems": false,
        "enabled": true,
        "isShowInPercentEnabled": false,
        "isShowInPercentVisible": false,
        "itemsLimit": 1,
        "title": "Metric",
      },
    },
    "customError": undefined,
    "exportConfig": {
      "supported": false,
    },
    "openAsReport": {
      "supported": false,
    },
    "recommendations": {
      "supportsFilteringRecommendation": true,
    },
    "supportedOverTimeComparisonTypes": [],
  },
}
`;
